.devcontainer/entrypoint.sh

```diff
@@ -69,6 +70,7 @@ ALLOWED_HOSTS = ['${API_HOST}', '*']
 ADMIN_URL = 'admin/'
 
 CORS_ORIGIN_ALLOW_ALL = True
+CORS_ORIGIN_WHITELIST = ['https://${APP_HOST}', 'https://${API_HOST}']
```

Comment:
does it still work if you remove this? the line right above it `CORS_ORIGIN_ALLOW_ALL = True` is so that we don't need this setting

Reply:
No. Without that entry, I get an `AttributeError: 'Settings' object has no attribute 'CORS_ORIGIN_WHITELIST'` error. Let me do a full rebuild to see if an empty array will work.

```text
trmm-api-dev          | HTTP OPTIONS /accounts/users/setup_totp/ 200 [0.02, 192.168.220.30:36946]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:01:36:05 +0000] "OPTIONS /accounts/users/setup_totp/ HTTP/1.1" 200 0 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-api-dev          | Internal Server Error: /accounts/users/setup_totp/
trmm-api-dev          | Traceback (most recent call last):
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 534, in thread_handler
trmm-api-dev          |     raise exc_info[1]
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/core/handlers/exception.py", line 42, in inner
trmm-api-dev          |     response = await get_response(request)
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 534, in thread_handler
trmm-api-dev          |     raise exc_info[1]
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
trmm-api-dev          |     response = await wrapped_callback(
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 479, in __call__
trmm-api-dev          |     ret: _R = await loop.run_in_executor(
trmm-api-dev          |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/current_thread_executor.py", line 40, in run
trmm-api-dev          |     result = self.fn(*self.args, **self.kwargs)
trmm-api-dev          |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 538, in thread_handler
trmm-api-dev          |     return func(*args, **kwargs)
trmm-api-dev          |            ^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
trmm-api-dev          |     return view_func(*args, **kwargs)
trmm-api-dev          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
trmm-api-dev          |     return self.dispatch(request, *args, **kwargs)
trmm-api-dev          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 509, in dispatch
trmm-api-dev          |     response = self.handle_exception(exc)
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 469, in handle_exception
trmm-api-dev          |     self.raise_uncaught_exception(exc)
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
trmm-api-dev          |     raise exc
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 506, in dispatch
trmm-api-dev          |     response = handler(request, *args, **kwargs)
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/accounts/views.py", line 205, in post
trmm-api-dev          |     return Response(TOTPSetupSerializer(user).data)
trmm-api-dev          |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/serializers.py", line 555, in data
trmm-api-dev          |     ret = super().data
trmm-api-dev          |           ^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/serializers.py", line 253, in data
trmm-api-dev          |     self._data = self.to_representation(self.instance)
trmm-api-dev          |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/serializers.py", line 522, in to_representation
trmm-api-dev          |     ret[field.field_name] = field.to_representation(attribute)
trmm-api-dev          |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/fields.py", line 1838, in to_representation
trmm-api-dev          |     return method(value)
trmm-api-dev          |            ^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/accounts/serializers.py", line 66, in get_qr_url
trmm-api-dev          |     obj.username, issuer_name=get_webdomain()
trmm-api-dev          |                               ^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/tacticalrmm/helpers.py", line 63, in get_webdomain
trmm-api-dev          |     return urlparse(settings.CORS_ORIGIN_WHITELIST[0]).netloc
trmm-api-dev          |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/conf/__init__.py", line 104, in __getattr__
trmm-api-dev          |     val = getattr(_wrapped, name)
trmm-api-dev          |           ^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          | AttributeError: 'Settings' object has no attribute 'CORS_ORIGIN_WHITELIST'
trmm-api-dev          | HTTP POST /accounts/users/setup_totp/ 500 [0.12, 192.168.220.30:36958]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:01:36:05 +0000] "POST /accounts/users/setup_totp/ HTTP/1.1" 500 121760 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
```

Reply:
With an empty array, I get an `IndexError: list index out of range` error.

```text
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:01:49:53 +0000] "GET /node_modules/.q-cache/vite/spa/deps/qrcode_vue.js?v=b97878ff HTTP/1.1" 200 29985 "https://rmm.a8n.tech/src/views/TOTPSetup.vue" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-api-dev          | HTTP OPTIONS /accounts/users/setup_totp/ 200 [0.06, 192.168.220.30:49220]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:01:49:54 +0000] "OPTIONS /accounts/users/setup_totp/ HTTP/1.1" 200 0 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-api-dev          | Internal Server Error: /accounts/users/setup_totp/
trmm-api-dev          | Traceback (most recent call last):
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 534, in thread_handler
trmm-api-dev          |     raise exc_info[1]
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/core/handlers/exception.py", line 42, in inner
trmm-api-dev          |     response = await get_response(request)
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 534, in thread_handler
trmm-api-dev          |     raise exc_info[1]
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
trmm-api-dev          |     response = await wrapped_callback(
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 479, in __call__
trmm-api-dev          |     ret: _R = await loop.run_in_executor(
trmm-api-dev          |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/current_thread_executor.py", line 40, in run
trmm-api-dev          |     result = self.fn(*self.args, **self.kwargs)
trmm-api-dev          |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/asgiref/sync.py", line 538, in thread_handler
trmm-api-dev          |     return func(*args, **kwargs)
trmm-api-dev          |            ^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
trmm-api-dev          |     return view_func(*args, **kwargs)
trmm-api-dev          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
trmm-api-dev          |     return self.dispatch(request, *args, **kwargs)
trmm-api-dev          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 509, in dispatch
trmm-api-dev          |     response = self.handle_exception(exc)
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 469, in handle_exception
trmm-api-dev          |     self.raise_uncaught_exception(exc)
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
trmm-api-dev          |     raise exc
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/views.py", line 506, in dispatch
trmm-api-dev          |     response = handler(request, *args, **kwargs)
trmm-api-dev          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/accounts/views.py", line 205, in post
trmm-api-dev          |     return Response(TOTPSetupSerializer(user).data)
trmm-api-dev          |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/serializers.py", line 555, in data
trmm-api-dev          |     ret = super().data
trmm-api-dev          |           ^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/serializers.py", line 253, in data
trmm-api-dev          |     self._data = self.to_representation(self.instance)
trmm-api-dev          |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/serializers.py", line 522, in to_representation
trmm-api-dev          |     ret[field.field_name] = field.to_representation(attribute)
trmm-api-dev          |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/env/lib/python3.11/site-packages/rest_framework/fields.py", line 1838, in to_representation
trmm-api-dev          |     return method(value)
trmm-api-dev          |            ^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/accounts/serializers.py", line 66, in get_qr_url
trmm-api-dev          |     obj.username, issuer_name=get_webdomain()
trmm-api-dev          |                               ^^^^^^^^^^^^^^^
trmm-api-dev          |   File "/workspace/api/tacticalrmm/tacticalrmm/helpers.py", line 63, in get_webdomain
trmm-api-dev          |     return urlparse(settings.CORS_ORIGIN_WHITELIST[0]).netloc
trmm-api-dev          |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
trmm-api-dev          | IndexError: list index out of range
trmm-api-dev          | HTTP POST /accounts/users/setup_totp/ 500 [0.23, 192.168.220.30:49226]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:01:49:54 +0000] "POST /accounts/users/setup_totp/ HTTP/1.1" 500 117213 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
```

Reply:
`CORS_ORIGIN_WHITELIST` is needed but it doesn't need the `'https://${API_HOST}'`. I'll remove that.

Here are the logs with `CORS_ORIGIN_WHITELIST = ['https://${APP_HOST}']`.
```text
trmm-api-dev          | HTTP OPTIONS /checkcreds/ 200 [0.05, 192.168.220.30:36952]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:02:06:00 +0000] "OPTIONS /checkcreds/ HTTP/1.1" 200 0 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-api-dev          | HTTP POST /checkcreds/ 200 [0.52, 192.168.220.30:36954]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:02:06:00 +0000] "POST /checkcreds/ HTTP/1.1" 200 137 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:02:06:00 +0000] "GET /src/views/TOTPSetup.vue HTTP/1.1" 200 16910 "https://rmm.a8n.tech/src/router/routes.js" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:02:06:00 +0000] "GET /node_modules/.q-cache/vite/spa/deps/qrcode_vue.js?v=7c17b797 HTTP/1.1" 200 30213 "https://rmm.a8n.tech/src/views/TOTPSetup.vue" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-api-dev          | HTTP OPTIONS /accounts/users/setup_totp/ 200 [0.02, 192.168.220.30:36970]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:02:06:00 +0000] "OPTIONS /accounts/users/setup_totp/ HTTP/1.1" 200 0 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
trmm-api-dev          | HTTP POST /accounts/users/setup_totp/ 200 [0.07, 192.168.220.30:36982]
trmm-nginx-dev        | 192.168.220.1 - - [14/Nov/2023:02:06:01 +0000] "POST /accounts/users/setup_totp/ HTTP/1.1" 200 177 "https://rmm.a8n.tech/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0" "-"
```